### ISSUE TITLE
tests doesn't work after merging #194

### ISSUE THREAD

**Issue by gilv** [pywren/pywren-ibm-cloud#197]
@JosepSampe i merged #194 
Now tests doesn't work...is this expected, because #194 is not backward compatible?

Thanks

======================================================================
ERROR: test_chunks_bucket (__main__.TestPywren)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "pywren_ibm_cloud/tests.py", line 324, in test_chunks_bucket
    pw.map_reduce(my_map_function_bucket, data_prefix, my_reduce_function, chunk_size=1 * 1024 ** 2)
  File "/Users/gilv/.pyenv/versions/3.6.7/lib/python3.6/site-packages/pywren_ibm_cloud/executor.py", line 213, in map_reduce
    execution_timeout=timeout)
  File "/Users/gilv/.pyenv/versions/3.6.7/lib/python3.6/site-packages/pywren_ibm_cloud/job/job.py", line 47, in create_map_job
    arg_data = utils.verify_args(map_function, data)
  File "/Users/gilv/.pyenv/versions/3.6.7/lib/python3.6/site-packages/pywren_ibm_cloud/utils.py", line 197, in verify_args
    new_elem = dict(new_func_sig.bind(elem).arguments)
  File "/Users/gilv/.pyenv/versions/3.6.7/lib/python3.6/inspect.py", line 2989, in bind
    return args[0]._bind(args[1:], kwargs)
  File "/Users/gilv/.pyenv/versions/3.6.7/lib/python3.6/inspect.py", line 2904, in _bind
    raise TypeError(msg) from None
TypeError: missing a required argument: 'key'

======================================================================
ERROR: test_chunks_bucket_one_reducer_per_object (__main__.TestPywren)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "pywren_ibm_cloud/tests.py", line 332, in test_chunks_bucket_one_reducer_per_object
    reducer_one_per_object=True)
  File "/Users/gilv/.pyenv/versions/3.6.7/lib/python3.6/site-packages/pywren_ibm_cloud/executor.py", line 213, in map_reduce
    execution_timeout=timeout)
  File "/Users/gilv/.pyenv/versions/3.6.7/lib/python3.6/site-packages/pywren_ibm_cloud/job/job.py", line 47, in create_map_job
    arg_data = utils.verify_args(map_function, data)
  File "/Users/gilv/.pyenv/versions/3.6.7/lib/python3.6/site-packages/pywren_ibm_cloud/utils.py", line 197, in verify_args
    new_elem = dict(new_func_sig.bind(elem).arguments)
  File "/Users/gilv/.pyenv/versions/3.6.7/lib/python3.6/inspect.py", line 2989, in bind
    return args[0]._bind(args[1:], kwargs)
  File "/Users/gilv/.pyenv/versions/3.6.7/lib/python3.6/inspect.py", line 2904, in _bind
    raise TypeError(msg) from None
TypeError: missing a required argument: 'key'

======================================================================
ERROR: test_cloudobject (__main__.TestPywren)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "pywren_ibm_cloud/tests.py", line 339, in test_cloudobject
    pw.map_reduce(my_cloudobject_put, data_prefix, my_cloudobject_get)
  File "/Users/gilv/.pyenv/versions/3.6.7/lib/python3.6/site-packages/pywren_ibm_cloud/executor.py", line 213, in map_reduce
    execution_timeout=timeout)
  File "/Users/gilv/.pyenv/versions/3.6.7/lib/python3.6/site-packages/pywren_ibm_cloud/job/job.py", line 47, in create_map_job
    arg_data = utils.verify_args(map_function, data)
  File "/Users/gilv/.pyenv/versions/3.6.7/lib/python3.6/site-packages/pywren_ibm_cloud/utils.py", line 197, in verify_args
    new_elem = dict(new_func_sig.bind(elem).arguments)
  File "/Users/gilv/.pyenv/versions/3.6.7/lib/python3.6/inspect.py", line 2989, in bind
    return args[0]._bind(args[1:], kwargs)
  File "/Users/gilv/.pyenv/versions/3.6.7/lib/python3.6/inspect.py", line 2904, in _bind
    raise TypeError(msg) from None
TypeError: missing a required argument: 'key'

======================================================================
ERROR: test_map_reduce_cos_bucket (__main__.TestPywren)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "pywren_ibm_cloud/tests.py", line 281, in test_map_reduce_cos_bucket
    pw.map_reduce(my_map_function_bucket, data_prefix, my_reduce_function)
  File "/Users/gilv/.pyenv/versions/3.6.7/lib/python3.6/site-packages/pywren_ibm_cloud/executor.py", line 213, in map_reduce
    execution_timeout=timeout)
  File "/Users/gilv/.pyenv/versions/3.6.7/lib/python3.6/site-packages/pywren_ibm_cloud/job/job.py", line 47, in create_map_job
    arg_data = utils.verify_args(map_function, data)
  File "/Users/gilv/.pyenv/versions/3.6.7/lib/python3.6/site-packages/pywren_ibm_cloud/utils.py", line 197, in verify_args
    new_elem = dict(new_func_sig.bind(elem).arguments)
  File "/Users/gilv/.pyenv/versions/3.6.7/lib/python3.6/inspect.py", line 2989, in bind
    return args[0]._bind(args[1:], kwargs)
  File "/Users/gilv/.pyenv/versions/3.6.7/lib/python3.6/inspect.py", line 2904, in _bind
    raise TypeError(msg) from None
TypeError: missing a required argument: 'key'

======================================================================
ERROR: test_map_reduce_cos_bucket_one_reducer_per_object (__main__.TestPywren)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "pywren_ibm_cloud/tests.py", line 288, in test_map_reduce_cos_bucket_one_reducer_per_object
    pw.map_reduce(my_map_function_bucket, data_prefix, my_reduce_function, reducer_one_per_object=True)
  File "/Users/gilv/.pyenv/versions/3.6.7/lib/python3.6/site-packages/pywren_ibm_cloud/executor.py", line 213, in map_reduce
    execution_timeout=timeout)
  File "/Users/gilv/.pyenv/versions/3.6.7/lib/python3.6/site-packages/pywren_ibm_cloud/job/job.py", line 47, in create_map_job
    arg_data = utils.verify_args(map_function, data)
  File "/Users/gilv/.pyenv/versions/3.6.7/lib/python3.6/site-packages/pywren_ibm_cloud/utils.py", line 197, in verify_args
    new_elem = dict(new_func_sig.bind(elem).arguments)
  File "/Users/gilv/.pyenv/versions/3.6.7/lib/python3.6/inspect.py", line 2989, in bind
    return args[0]._bind(args[1:], kwargs)
  File "/Users/gilv/.pyenv/versions/3.6.7/lib/python3.6/inspect.py", line 2904, in _bind
    raise TypeError(msg) from None
TypeError: missing a required argument: 'key'

======================================================================
ERROR: test_map_reduce_cos_key (__main__.TestPywren)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "pywren_ibm_cloud/tests.py", line 294, in test_map_reduce_cos_key
    iterdata = [bucket_name + '/' + key for key in list_test_keys()]
  File "pywren_ibm_cloud/tests.py", line 53, in list_test_keys
    return STORAGE.list_keys_with_prefix(bucket_name=STORAGE_CONFIG['bucket'],
AttributeError: 'StorageBackend' object has no attribute 'list_keys_with_prefix'

======================================================================
ERROR: test_map_reduce_cos_key_one_reducer_per_object (__main__.TestPywren)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "pywren_ibm_cloud/tests.py", line 302, in test_map_reduce_cos_key_one_reducer_per_object
    iterdata = [bucket_name + '/' + key for key in list_test_keys()]
  File "pywren_ibm_cloud/tests.py", line 53, in list_test_keys
    return STORAGE.list_keys_with_prefix(bucket_name=STORAGE_CONFIG['bucket'],
AttributeError: 'StorageBackend' object has no attribute 'list_keys_with_prefix'

======================================================================
ERROR: test_map_reduce_url (__main__.TestPywren)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "pywren_ibm_cloud/tests.py", line 310, in test_map_reduce_url
    pw.map_reduce(my_map_function_url, TEST_FILES_URLS, my_reduce_function)
  File "/Users/gilv/.pyenv/versions/3.6.7/lib/python3.6/site-packages/pywren_ibm_cloud/executor.py", line 213, in map_reduce
    execution_timeout=timeout)
  File "/Users/gilv/.pyenv/versions/3.6.7/lib/python3.6/site-packages/pywren_ibm_cloud/job/job.py", line 47, in create_map_job
    arg_data = utils.verify_args(map_function, data)
  File "/Users/gilv/.pyenv/versions/3.6.7/lib/python3.6/site-packages/pywren_ibm_cloud/utils.py", line 197, in verify_args
    new_elem = dict(new_func_sig.bind(elem).arguments)
  File "/Users/gilv/.pyenv/versions/3.6.7/lib/python3.6/inspect.py", line 2989, in bind
    return args[0]._bind(args[1:], kwargs)
  File "/Users/gilv/.pyenv/versions/3.6.7/lib/python3.6/inspect.py", line 2904, in _bind
    raise TypeError(msg) from None
TypeError: missing a required argument: 'data_stream'

======================================================================
ERROR: test_storage_handler (__main__.TestPywren)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "pywren_ibm_cloud/tests.py", line 315, in test_storage_handler
    iterdata = [[key, STORAGE_CONFIG['bucket']] for key in list_test_keys()]
  File "pywren_ibm_cloud/tests.py", line 53, in list_test_keys
    return STORAGE.list_keys_with_prefix(bucket_name=STORAGE_CONFIG['bucket'],
AttributeError: 'StorageBackend' object has no attribute 'list_keys_with_prefix'

----------------------------------------------------------------------
Ran 14 tests in 164.045s

FAILED (errors=9)
Deleting test files...
Traceback (most recent call last):
  File "pywren_ibm_cloud/tests.py", line 380, in <module>
    cleanTests()
  File "pywren_ibm_cloud/tests.py", line 59, in cleanTests
    for key in list_test_keys():
  File "pywren_ibm_cloud/tests.py", line 53, in list_test_keys
    return STORAGE.list_keys_with_prefix(bucket_name=STORAGE_CONFIG['bucket'],
AttributeError: 'StorageBackend' object has no attribute 'list_keys_with_prefix'
gilv1-mac:pywren-ibm-cloud gilv$

**Comment by gilv** [comment]
@JosepSampe it's a salad of failures ;)

1. TypeError: missing a required argument: 'key'
2. AttributeError: 'StorageBackend' object has no attribute 'list_keys_with_prefix'
3. TypeError: missing a required argument: 'data_stream'

**Comment by JosepSampe** [comment]
@gil yesterday I already asked @omer in #192 to adapt tests to the new changes since he knows better than me how tests flow works

**Comment by gilv** [comment]
@JosepSampe my question is more general...the "obj" and "url" adittions, are not backward compatible, right? We need to change all the code that used, is this correct?

**Comment by gilv** [comment]
@JosepSampe does "obj" and "cloudobject" can co-exists together?

**Comment by JosepSampe** [comment]
> @JosepSampe my question is more general...the "obj" and "url" adittions, are not backward compatible, right? We need to change all the code that used, is this correct?

No, it is not backward compatible. Internal logic is completely different with 'obj' and it can't coexists with older args.


> @JosepSampe does "obj" and "cloudobject" can co-exists together?

In fact 'obj' is a CloudObject with extra parameters ;)

**Comment by gilv** [comment]
@JosepSampe  so get_results also return lists of those new "obj"s right?

**Comment by JosepSampe** [comment]
@gilv hmmm.. no, get_result() returns what you return in the return statement of the function. 'obj' is only an input arg.

**Comment by gilv** [comment]
@JosepSampe but today ret_result can also returns cloudobjects...this is what we had already. Lachlan asked for it.. Then i am confused how it works now, if obj is extended cloudobject so what is returned in get_results?

def my_cloudobject_get(results, internal_storage):
    data = [internal_storage.get_object(cloudobject) for cloudobject in results]
    return my_reduce_function(data)

**Comment by gilv** [comment]
@JosepSampe i actually even more confused now..
https://github.com/pywren/pywren-ibm-cloud/pull/169

How it all fits with "obj" ?

**Comment by JosepSampe** [comment]
@gilv cloud object is simply a class that contains bucket and key:
```
class CloudObject:
    _init__(self, bucket, key):
         self.bucket = bucket
         self.key = key
```

So when a user writes a cos:// url, pywren maps each object in COS to one instance of this class.

internal_storage.put_object is another thing. It is a method to put temp objects in COS. To do so it internally uses instances of CloudObject.

From a function activation you can return whatever data type you want, cloudobjects, lists, dicts, ints, strings, other classes, etc etc... When you process data from COS, you only need to return the result as we always do, why you want to return cloudObjects here?

**Comment by gilv** [comment]
@JosepSampe in #169 you already added support to return cloudobjects...

`````
import pywren_ibm_cloud as pywren

def my_function_put(text, internal_storage):
    co1 = internal_storage.put_object('Temp object test 1: {}'.format(text, ))
    co2 = internal_storage.put_object('Temp object test 2: {}'.format(text, ))

    return [co1, co2]

def my_function_get(co, internal_storage):
    data = internal_storage.get_object(co)

    return data

pw = pywren.ibm_cf_executor()
pw.call_async(my_function_put, 'Hello World')
cloudobjects = pw.get_result()
pw.map(my_function_get, cloudobjects)
print(pw.get_result())
pw.clean()
``````

will this continue to work?

**Comment by JosepSampe** [comment]
@gilv Yeap, as I said in the previous post you can return whatever you want ;)

**Comment by JosepSampe** [comment]
@gilv All tests fixed and running ok in #198
